gms/gossiper.cc

```diff
@@ -421,6 +425,120 @@ future<> gossiper::handle_echo_msg() {
+        auto state = get_application_state_ptr(node, application_state::TOKENS);
+        auto tokens = state ? state->value : "";
+        ser::serialize(out, tokens);
+        ser::serialize(out, get_gossip_status(node));
+        ser::serialize(out, node);
+        feed_hash(h, out.linearize());
+    }
+    return h.finalize_uint64();
+}
+
+static future<cluster_ring_status_response>
+get_cluster_ring_status(gossiper& g) {
+    using response = cluster_ring_status_response;
+    return g.container().map([] (gossiper& g) -> future<response> {
+        return seastar::async([&g] {
+            auto& tm = service::get_local_storage_service().get_token_metadata();
+            cluster_ring_status_response resp = {{
+                {response::nr_known_nodes, format("{}", g.endpoint_state_map.size())},
+                {response::status_token_hash, format("{}", g.get_gossip_status_token_hash())},
+                {response::normal_token_hash, format("{}", tm.get_normal_token_hash())},
+                {response::bootstrap_token_hash, format("{}", tm.get_bootstrap_token_hash())},
+                {response::leaving_nodes_hash, format("{}", tm.get_leaving_endpoints_hash())},
+                {response::pending_ranges_hash, format("{}", tm.get_pending_ranges_hash())}
+            }};
+            return resp;
+        });
+    }).then([] (std::vector<cluster_ring_status_response> results) {
+        bool same_results = std::adjacent_find(results.begin(), results.end(),
+                std::not_equal_to<cluster_ring_status_response>()) == results.end();
+        results.front().status[response::same_on_all_shards] = same_results ?
+                response::same_on_all_shards_yes : response::same_on_all_shards_no;
+        return results.front();
+    });
+}
+
+// Send CLUSTER_RING_STATUS RPC verb to remote nodes to query the cluster
+// status. If local node has the same cluster ring status as the remote nodes,
+// it means the cluster has the same view on the cluster ring status.
+static future<> wait_for_cluster_ring_status_to_match(gossiper& g) {
+    return seastar::async([&g] {
+        while (true) {
+            auto local = get_cluster_ring_status(g).get0();
+            auto myip = g.get_broadcast_address();
+            if (!local.is_same_on_all_shards()) {
+                logger.debug("Node {}, status = {}, wait for all the shards have the same hashes", myip, local.status);
+                sleep_abortable(std::chrono::milliseconds(100)).get();
+                continue;
+            }
+            auto nodes_to_check = boost::copy_range<std::set<gms::inet_address>>(g.endpoint_state_map | boost::adaptors::map_keys);
+            for (auto& seed : g.get_seeds()) {
+                nodes_to_check.insert(seed);
+            }
+            nodes_to_check.erase(myip);
+            auto node_to_be_replaced = g.get_node_to_be_replaced();
+            if (node_to_be_replaced) {
+                nodes_to_check.erase(*node_to_be_replaced);
```

Comment:
We probably can, also for the node that is removed by nodetool removenode. 

Consider this: the booting node think a node is in LEFT status but this information is out of date. A node with the same IP has bootstrapped again into the cluster. The booting node will wrongly ignore sending query to that node.

However, it does not matter, because the hash will not match in this round anyway. In next round, the status of the node in question becomes the correct state.

For SHUTDOWN status, I am not sure if we should skip it or not. We probably can.

Reply:
I wouldn't skip SHUTDOWN because it's the same case as with a node which does not respond.

Reply:
SHUTDOWN is a node shutdown itself and announces that. In most cases, it is is really down. The rpc verb would endup with seastar::rpc::closed_error so we know this node is down. Currently, we fallback to the old method if there is a node down, because the down node can not reply a hash,  we can wait forever for no good reason.

Reply:
I don't know if you agree or disagree about SHUTDOWN.

Reply:
I disagree it is the same case as a node does not respond in  "I wouldn't skip SHUTDOWN because it's the same case as with a node which does not respond".

A shutdown should faill the rpc request attempt immediately. 

I do not want to filter out SHUTDOWN for now in this version. This means we fallback to the old method.

Reply:
Ok